influxql/engine.go

```diff
@@ -349,6 +359,126 @@ func (m *MapReduceJob) processRawQuery(out chan *Row, filterEmptyResults bool) {
 	}
 }
 
+// derivativeInterval returns the time interval for the one (and only) derivative func
+func (m *MapReduceJob) derivativeInterval() time.Duration {
+	return m.stmt.FunctionCalls()[0].Args[1].(*DurationLiteral).Val
+}
+
+func (m *MapReduceJob) isNonNegativeDerivative() bool {
+	return m.stmt.FunctionCalls()[0].Name == "non_negative_derivative"
+}
+
+func (m *MapReduceJob) processRawQueryDerivative(lastValueFromPreviousChunk *rawQueryMapOutput, valuesToReturn []*rawQueryMapOutput) []*rawQueryMapOutput {
+	// If we're called and do not have a derivative aggregate function, then return what was passed in
+	if !m.stmt.HasDerivative() {
+		return valuesToReturn
+	}
+
+	if len(valuesToReturn) == 0 {
+		return valuesToReturn
+	}
+
+	// If we only have 1 value, then the value did not change, so return
+	// a single row w/ 0.0
+	if len(valuesToReturn) == 1 {
+		return []*rawQueryMapOutput{
+			&rawQueryMapOutput{
+				Time:   valuesToReturn[0].Time,
+				Values: 0.0,
+			},
+		}
+	}
+
+	if lastValueFromPreviousChunk == nil {
+		lastValueFromPreviousChunk = valuesToReturn[0]
+	}
+
+	// Determines whether to drop negative differences
+	isNonNegative := m.isNonNegativeDerivative()
+
+	derivativeValues := []*rawQueryMapOutput{}
+	for i := 1; i < len(valuesToReturn); i++ {
+		v := valuesToReturn[i]
+
+		// Calculate the derivative of successive points by dividing the difference
+		// of each value by the elapsed time normalized to the interval
+		var value interface{}
+		diff := v.Values.(float64) - lastValueFromPreviousChunk.Values.(float64)
```

Comment:
You should only be using the last value from the previous chunk for the first value in this chunk. Otherwise, you should be looking at the previous point.